ADR-92 & AP-19: BFF proposal (#99)
bf2fc6cc711aee1a0c2a/architecture: _ap/19/index.adoc

---
num: 19
title: "Toward a more unified UX: BFF"
status: "Draft"
authors:
  - "Andrea Peruffo"
  - "Riccardo Forina"
  - "Eric Wittmann"
tags: []
---

## Intent

Define an architectural component to better decouple backend services and frontend needs offering a more cohesive UX across devices and tools (such as UI and CLI).

## Motivation

UI and CLI needs to poll information from multiple separate sources causing:

* poor UX, there is need for multiple network calls to offer simple functionalities
* code duplication, data aggregation logic is scattered and duplicated across different users
* testing and mocking, decoupling frontend and backend systems with only one component

## Applicability

This pattern applies when the functionality offered by frontend/CLI/etc. needs to access various APIs and do data-aggregation logic in various forms.
In those cases placing a frontend opinionated component in front can be an appropriate solution to speed up development and be a single source of truth for integration logic.

## Description

> Implementing a https://samnewman.io/patterns/architectural/bff/[BFF] for the backend services to better decouple and centralize the integration logic.

Reviewer: Add the `propagate the access token` concept in the description.
Mention the alternative being GraphQL but that for the team, expertise was not deems high enough to explore that path. This will leave the doors for other teams to explore.

Author: 1. done
2. I do believe that the current wording reflect weel enough our position toward GraphQL: https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/99/commits/877e4c8dac5be9ea3b2736bb0c73004adea04531

Reviewer: On 2, it's in the ADR and not in the AP, It is worth repeating to help consumption of the materials.

Author: Any reference to the implementing technology has been deliberately omitted from the AP, unless you have a really strong opinion I would keep it this way.

Reviewer: Yes I think it's important. It's not an implementation detail to go for a custom REST API vs a GraphQL based API, this is a defining choice.

Author: > this is a defining choice

can you please elaborate?
This interpretation is completely alien to me and, I believe, to anyone else involved in drafting this proposal.
cc. @EricWittmann @riccardo-forina

Reviewer 2: I'm also for not mentioning any particular technology here. A BFF could be implemented in a number of ways and there is no "best" or "worst" technology, but a "right" one that fully depends on the context.

For example, for a team that's using Typescript for the full stack and fully control the clients connecting to the backend, tRPC is a great choice, arguably better than GraphQL or plain old REST+OpenAPI

Author: @emmanuelbernard if you are convinced that this aspect is important for the AP I'm happy to remove the AP from this PR and let you(or someone else) take over in another iteration.
In this case, I disagree with mixing scopes:
- BFF is the Architectural Pattern we want to promote as a possible solution to common issues
- OpenAPI/GraphQL is the technical instrument used to implement it, which is a completely orthogonal discussion - i.e. it can be implemented with CORBA or any other technology that serves the purpose

We briefly touched on this point with @chirino and I felt alignment on this aspect.

Reviewer 3: Lol you mention COBRA in the same breath as GraphQL as being fitting for a BFF pattern.  I’m not sure you understood our conversation at all.

Reviewer: I am proposing changes as part of a new commit. See https://github.com/andreaTP/architecture/pull/1

Author: Thanks @emmanuelbernard , the context you proposed is fine for me, merged 👍 

@chirino I firmly believe that CORBA is a perfectly valid fit for an application in COBOL ( [ref](http://computer-programming-forum.com/48-cobol/f4fdc89fc2d9bfb3.htm) )

In those cases placing a frontend opinionated component in front can be an appropriate solution to speed up development and be a single source of truth for integration logic.

A single BFF component is planned to host the API interactions of our various application services (Kafka, Connectors, Service registry etc).
This is a simplification choice that could be revisited if it becomes burdensome.
Until it is, alignment is expected.

## Description

Implementing a https://samnewman.io/patterns/architectural/bff/[BFF] for the backend services to better decouple and centralize the integration logic.

image::with_BFF.png[Proposed Architecture]

Following are key contextual information to understand the pattern and its potential limitations.

Access to the underlying services will be done by propagating the original (user) access token to simplify development and to keep the security model similar to the current architecture.
See ADR 91 for more details.

The implementation choice went for a custom REST endpoint implementation as opposed to using GraphQL.
GraphQL is a popular technology built to address this or a similar set of use cases.

> The team's experience lead tot he choice of a custom REST endpoint but could be revisited by this or another team.

Reviewer 4: ```suggestion
The team's experience led to the choice of a custom REST endpoint but could be revisited by this or another team.
```
